sync/src/status.rs

```diff
@@ -0,0 +1,187 @@
+use crate::{BAD_MESSAGE_BAN_TIME, SYNC_USELESS_BAN_TIME};
+use std::fmt::{self, Display, Formatter};
+use std::time::Duration;
+
+/// Similar to `?`, `attempt!` is used for propagating `Status`.
+///
+/// `attempt!` return early if it is not `Status::ok()`.
+///
+/// ```rust
+/// use ckb_sync::{Status, StatusCode, attempt};
+///
+/// fn return_early(status: Status) -> Status {
+///     attempt!(status);
+///     StatusCode::OK.with_context("bar")
+/// }
+///
+/// assert_eq!(return_early(StatusCode::OK.into()).to_string(), "OK(10000): bar");
+/// assert_eq!(return_early(StatusCode::Ignored.into()).to_string(), "Ignored(10001)");
+/// ```
+#[macro_export]
+macro_rules! attempt {
+    ($code:expr) => {{
+        let ret = $code;
+        if !ret.is_ok() {
+            return ret;
+        }
+        ret
+    }};
+}
+
+/// StatusCodes indicate whether a specific operation has been successfully completed.
+/// The StatusCode element is a 5-digit integer.
+///
+/// The first digest of the StatusCode defines the class of result:
+///   - 1xxxx: Informational - Request received, continuing process
+///   - 4xxxx: Malformed errors - The request contains malformed messages
+///   - 5xxxx: Node errors - The node internal failed
+///
+/// The 2nd-3rd digests of the StatusCode defines the scope of result:
+///   - x00xx: General, undefined specific scope
+///   - x01xx: CompactBlock
+///   - x02xx: RelayTransactions
+///   - x03xx: RelayTransactionHashes
+///   - x04xx: GetRelayTransactions
+///   - x05xx: GetBlockTransactions
+///   - x06xx: BlockTransactions
+///   - x07xx: GetBlockProposal
+///   - x08xx: BlockProposal
+///   - x09xx: GetHeaders
+///   - x10xx: SendHeaders
+///   - x11xx: GetBlocks
+///   - x12xx: SendBlock
+///   - x13xx: InIBD
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum StatusCode {
```

Comment:
Under the default representation, the specified discriminant is interpreted as an isize value although the compiler is allowed to use a smaller type in the actual memory layout. The size and thus acceptable values can be changed by using a primitive representation.
`#[repr(u16)]`, suggestion